src/app/EventManagement.cpp

```diff
@@ -0,0 +1,867 @@
+    VerifyOrExit(aEventOptions.mpEventSchema != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    err = LogEventPrivate(apDelegate, aEventOptions, aEventNumber);
+
+exit:
+    Platform::CriticalSectionExit();
+    ChipLogFunctError(err);
+    return err;
+}
+
+CHIP_ERROR EventManagement::LogEventPrivate(EventLoggingDelegate * apDelegate, EventOptions & aEventOptions,
+                                            EventNumber & aEventNumber)
+{
+    CircularTLVWriter writer;
+    CHIP_ERROR err                 = CHIP_NO_ERROR;
+    uint32_t requestSize           = 0;
+    aEventNumber                   = 0;
+    CircularEventBuffer checkpoint = *mpEventBuffer;
+    CircularEventBuffer * buffer   = nullptr;
+    EventLoadOutContext ctxt       = EventLoadOutContext(writer, aEventOptions.mpEventSchema->mPriority,
+                                                   GetPriorityBuffer(aEventOptions.mpEventSchema->mPriority)->mLastEventNumber);
+    Timestamp timestamp(Timestamp::Type::kSystem, System::Timer::GetCurrentEpoch());
+    EventOptions opts = EventOptions(timestamp);
+    // Start the event container (anonymous structure) in the circular buffer
+    writer.Init(*mpEventBuffer);
+
+    // check whether the entry is to be logged or discarded silently
+    VerifyOrExit(aEventOptions.mpEventSchema->mPriority >= CHIP_CONFIG_EVENT_GLOBAL_PRIORITY, /* no-op */);
+
+    // Create all event specific data
+    // Timestamp; encoded as a delta time
+    if (aEventOptions.mTimestamp.mType == Timestamp::Type::kSystem)
+    {
+        opts.mTimestamp = aEventOptions.mTimestamp;
+    }
+
+    if (GetPriorityBuffer(aEventOptions.mpEventSchema->mPriority)->mFirstEventSystemTimestamp.mValue == 0)
+    {
+        GetPriorityBuffer(aEventOptions.mpEventSchema->mPriority)->UpdateFirstLastEventTime(opts.mTimestamp);
+    }
+
+    opts.mUrgent       = aEventOptions.mUrgent;
+    opts.mpEventSchema = aEventOptions.mpEventSchema;
+
+    ctxt.mCurrentEventNumber = GetPriorityBuffer(opts.mpEventSchema->mPriority)->mLastEventNumber;
+    ctxt.mCurrentSystemTime  = GetPriorityBuffer(opts.mpEventSchema->mPriority)->mLastEventSystemTimestamp;
+
+    err = CalculateEventSize(apDelegate, &opts, requestSize);
+    SuccessOrExit(err);
+
+    // Ensure we have space in the in-memory logging queues
+    err = EnsureSpaceInCircularBuffer(requestSize);
+    SuccessOrExit(err);
+
+    err = ConstructEvent(&ctxt, apDelegate, &opts);
+    SuccessOrExit(err);
+
+    // Check the number of bytes written.  If the event is too large
+    // to be evicted from subsequent buffers, drop it now.
+    buffer = mpEventBuffer;
+    do
+    {
+        VerifyOrExit(buffer->GetQueueSize() >= writer.GetLengthWritten(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        if (buffer->IsFinalDestinationForPriority(opts.mpEventSchema->mPriority))
+        {
+            break;
+        }
+        else
+        {
+            buffer = buffer->GetNextCircularEventBuffer();
```

Comment:
Add an assert that buffer is not nullptr.


consider changing this to `while(true)` instead of `do { } while(true)` to have only one form of looping for inifinte loops in the code.